File: frontend/src/api/schedule.ts
import type { AgentRecord } from './agents.ts'
import { errorDetail } from './hierarchy.ts'

export type ActivityType = 'On Phone' | 'Paid Break' | 'Unpaid Meal' | 'Leave' | 'Training'
export type Activity = {
  activity_type: ActivityType
  start: string // ISO with offset, e.g. 2026-09-23T10:00:00-04:00
  end: string
  paid: boolean
  minutes?: number
  queue_id?: string | null
}
export type DayStatus = 'Working' | 'Off' | 'Leave' | 'Unscheduled' | 'Outside horizon'

// --- Agent calendar (real plan, or the labelled mock in the same shape) --------------------

export type PlanDay = {
  date: string
  weekday: string
  status: DayStatus
  off: boolean
  shift_code?: string | null
  shift_name?: string | null
  start?: string | null
  end?: string | null
  paid_hours?: number
  preferred_shift_matched?: boolean | null
  preferred_day_off?: boolean | null
  note?: string | null
  activities?: Activity[]
}

export type PlanWeek = {
  week_start: string
  paid_hours: number
  days_off: number
  leave_days: number
  preferred_shift_matches: number | null
  working_days: number
  partial: boolean
  days: PlanDay[]
}

export type AgentPlan = {
  source: 'optimized' | 'mock'
  run_id: string | null
  validation_status: string
  queue_status: string | null
  timezone: string
  weeks: PlanWeek[]
}

export type AgentScheduleResponse = { agent: AgentRecord; schedule: AgentPlan }

export async function fetchAgentSchedule(agentId: string): Promise<AgentScheduleResponse> {
  const response = await fetch(`/api/agents/${encodeURIComponent(agentId)}/schedule`)
  if (!response.ok) throw new Error(await errorDetail(response, 'Failed to load schedule'))
  return (await response.json()) as AgentScheduleResponse
}

// --- Runs ---------------------------------------------------------------------------------

export type RunState = 'not_generated' | 'queued' | 'running' | 'completed' | 'completed_with_shortages' | 'failed'

export type StageResult = { stage: string; status: string; objective: number | null; bound: number | null; seconds: number }

export type RunSummary = {
  run_id: string
  created_at: string
  horizon: { start: string; end: string; days: number }
  status_counts: Record<string, number>
  shortage_agent_hours: number
  validation_passed: boolean
  runtime_seconds: number
  conflicts: string[]
  excluded_agents: { agent_id: string; queue_id: string; reason: string }[]
  config: { staffing: StaffingConfig; scheduling: SchedulingConfig }
  assumptions: string[]
}

export type LatestRun = {
  state: RunState
  run_id?: string
  message?: string
  done?: number
  total?: number
  latest_completed_run_id?: string | null
  summary?: RunSummary
  stale?: boolean
}

export type StaffingConfig = {
  interval_minutes: number
  service_level_target: number
  answer_threshold_seconds: number
  max_occupancy: number
  residual_shrinkage: number
  closing_minutes: number
  within_hour_proportions: number[] | null
}

export type SchedulingConfig = {
  break_stagger_slots: number
  stage_time_limit_seconds: number[]
  num_workers: number
  leave_csv: string | null
  assume_rested_at_start: boolean
  weights: Record<string, number>
}

export async function fetchLatestRun(): Promise<LatestRun> {
  const response = await fetch('/api/schedule/runs/latest')
  if (!response.ok) throw new Error(await errorDetail(response, 'Failed to load schedule status'))
  return (await response.json()) as LatestRun
}

export async function fetchScheduleConfig(): Promise<{ staffing: StaffingConfig; scheduling: SchedulingConfig }> {
  const response = await fetch('/api/schedule/config')
  if (!response.ok) throw new Error(await errorDetail(response, 'Failed to load schedule settings'))
  return (await response.json()) as { staffing: StaffingConfig; scheduling: SchedulingConfig }
}

export async function startRun(overrides: {
  staffing: Partial<StaffingConfig>
  scheduling: Partial<SchedulingConfig>
}): Promise<{ run_id: string }> {
  const response = await fetch('/api/schedule/runs', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(overrides),
  })
  if (!response.ok) throw new Error(await errorDetail(response, 'Failed to start schedule run'))
  return (await response.json()) as { run_id: string }
}

// --- Queue schedule -------------------------------------------------------------------------

export type CoverageRow = {
  queue_id: string
  interval_start: string
  interval_end: string
  interval_minutes: number
  kind: 'open' | 'closing' | 'no demand'
  required_on_phone: number
  scheduled_on_phone: number
  shortage: number
  excess: number
}

export type ScheduleDay = {
  work_date: string
  status: DayStatus
  shift_code: string | null
  shift_name: string | null
  shift_start: string | null
  shift_end: string | null
  paid_minutes: number
  unpaid_minutes: number
  preferred_shift_matched: boolean | null
  preferred_day_off: boolean | null
  note: string | null
  activities: Activity[]
}

export type ScheduleAgent = {
  agent_id: string
  agent_name: string
  paid_minutes: number
  working_days: number
  preferred_shift_matches: number
  days: ScheduleDay[]
}

export type Totals = {
  required_on_phone_hours: number
  scheduled_on_phone_hours: number
  shortage_agent_hours: number
  excess_agent_hours: number
  paid_hours: number
  leave_hours: number
  working_shifts: number
  agents: number
  peak_required_on_phone: number
}

export type QueueSchedule = {
  run_id: string
  created_at: string
  horizon: { start: string; end: string; days: number }
  stale: boolean
  validation_passed: boolean
  queue: { queue_id: string; status: string; stages: StageResult[]; shortage_agent_hours?: number }
  totals: Totals
  coverage: CoverageRow[]
  agents: ScheduleAgent[]
}

export async function fetchQueueSchedule(queueId: string): Promise<QueueSchedule> {
  const response = await fetch(`/api/queues/${encodeURIComponent(queueId)}/schedule`)
  if (!response.ok) throw new Error(await errorDetail(response, 'Failed to load schedule'))
  return (await response.json()) as QueueSchedule
}

export type ScheduleSummary = {
  run_id: string
  level: 'org' | 'bu' | 'mu'
  code: string
  children: (Totals & { code: string })[]
  total: Totals
}

export async function fetchScheduleSummary(level: 'org' | 'bu' | 'mu', code: string): Promise<ScheduleSummary> {
  const params = new URLSearchParams({ level, code })
  const response = await fetch(`/api/schedule/summary?${params}`)
  if (!response.ok) throw new Error(await errorDetail(response, 'Failed to load schedule summary'))
  return (await response.json()) as ScheduleSummary
}

export function exportUrl(runId: string, kind: 'shifts' | 'activities' | 'coverage' | 'requirements', queueId?: string): string {
  const q = queueId ? `?queue_id=${encodeURIComponent(queueId)}` : ''
  return `/api/schedule/runs/${encodeURIComponent(runId)}/export/${kind}.csv${q}`
}

File: backend/src/wfm/api/schedule.py
"""Schedule endpoints: start a run, poll it, read queue/agent plans, roll-ups and CSV exports.

Runs execute in a worker subprocess (one at a time); results are read from the run folder.
"""

import io
import json
import subprocess
import sys
from dataclasses import dataclass
from datetime import date, timedelta
from functools import lru_cache
from pathlib import Path
from typing import Annotated, Any, Literal

import pandas as pd
from fastapi import APIRouter, HTTPException, Query
from fastapi.responses import StreamingResponse
from pydantic import BaseModel, ValidationError

from wfm.config import load_config, resolve
from wfm.forecast.store import file_fingerprint
from wfm.schedule.run import apply_overrides
from wfm.schedule.store import (
    active_run,
    latest_run,
    new_run_dir,
    read_status,
    schedules_root,
    write_status,
)

router = APIRouter(prefix="/api")

WEEKDAYS = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]


def _runs_dir() -> Path:
    return resolve(load_config().paths.runs_dir)


# --- Runs -------------------------------------------------------------------------------


class RunRequest(BaseModel):
    staffing: dict[str, Any] = {}
    scheduling: dict[str, Any] = {}


@router.get("/schedule/config")
def schedule_config() -> dict[str, Any]:
    config = load_config()
    return {
        "staffing": config.staffing.model_dump(mode="json"),
        "scheduling": config.scheduling.model_dump(mode="json"),
    }


@router.post("/schedule/runs", status_code=202)
def start_run(request: RunRequest) -> dict[str, Any]:
    overrides = request.model_dump()
    try:
        apply_overrides(load_config(), overrides)  # validate before starting anything
    except ValidationError as e:
        raise HTTPException(
            status_code=422, detail=f"Invalid settings: {e.errors()[0]['msg']}"
        ) from e
    runs_dir = _runs_dir()
    if running := active_run(runs_dir):
        raise HTTPException(status_code=409, detail=f"Run {running.name} is still in progress")
    run_dir = new_run_dir(runs_dir)
    write_status(
        run_dir,
        {"run_id": run_dir.name, "state": "queued", "message": "Starting", "done": 0, "total": 1},
    )
    with (run_dir / "worker.log").open("w") as log:
        process = subprocess.Popen(  # noqa: S603 — fixed interpreter and module, no shell
            [
                sys.executable,
                "-m",
                "wfm.schedule.run",
                "--run-dir",
                str(run_dir),
                "--overrides",
                json.dumps(overrides),
            ],
            stdout=log,
            stderr=subprocess.STDOUT,
            start_new_session=True,
        )
    write_status(
        run_dir,
        {
            "run_id": run_dir.name,
            "pid": process.pid,
            "state": "queued",
            "message": "Starting",
            "done": 0,
            "total": 1,
        },
    )
    return {"run_id": run_dir.name}


@router.get("/schedule/runs/latest")
def latest() -> dict[str, Any]:
    runs_dir = _runs_dir()
    run_dir = latest_run(runs_dir)
    if run_dir is None:
        return {"state": "not_generated"}
    status = read_status(run_dir) or {}
    completed = latest_run(runs_dir, completed_only=True)
    body: dict[str, Any] = {
        **status,
        "latest_completed_run_id": completed.name if completed else None,
    }
    if completed:
        meta = _load(completed.name).meta
        body["summary"] = _run_summary(meta)
        body["stale"] = _is_stale(meta)
    return body


def _run_summary(meta: dict[str, Any]) -> dict[str, Any]:
    return {
        "run_id": meta["run_id"],
        "created_at": meta["created_at"],
        "horizon": meta["horizon"],
        "status_counts": meta["status_counts"],
        "shortage_agent_hours": meta["shortage_agent_hours"],
        "validation_passed": meta["validation_passed"],
        "runtime_seconds": meta["runtime_seconds"],
        "conflicts": meta["conflicts"],
        "excluded_agents": meta["agents"]["excluded"],
        "config": meta["config"],
        "assumptions": meta["assumptions"],
    }


def _is_stale(meta: dict[str, Any]) -> bool:
    """True when the roster or ACD file changed since the run.

    Runs saved before fingerprints were recorded are treated as not stale (unknown).
    """
    inputs = meta.get("inputs")
    if not inputs:
        return False
    paths = load_config().paths
    current = {
        "agents_sha256": file_fingerprint(resolve(paths.agents_xlsx)),
        "acd_sha256": file_fingerprint(resolve(paths.acd_xlsx)),
    }
    return any(inputs.get(k) != v for k, v in current.items())


# --- Loading run outputs -----------------------------------------------------------------


@dataclass(frozen=True)
class RunData:
    meta: dict[str, Any]
    validation: dict[str, Any]
    shifts: pd.DataFrame
    activities: pd.DataFrame
    coverage: pd.DataFrame
    requirements: pd.DataFrame


@lru_cache(maxsize=2)
def _load(run_id: str) -> RunData:
    run_dir = schedules_root(_runs_dir()) / run_id
    if not (run_dir / "meta.json").is_file():
        raise HTTPException(status_code=404, detail=f"Schedule run {run_id} has no results")

    def read(name: str) -> pd.DataFrame:
        return pd.read_csv(run_dir / f"{name}.csv", keep_default_na=False, na_values=[""])

    return RunData(
        meta=json.loads((run_dir / "meta.json").read_text()),
        validation=json.loads((run_dir / "validation.json").read_text()),
        shifts=read("shifts"),
        activities=read("activities"),
        coverage=read("coverage"),
        requirements=read("requirements"),
    )


def _completed() -> RunData | None:
    run_dir = latest_run(_runs_dir(), completed_only=True)
    return _load(run_dir.name) if run_dir else None


def _clean(value: Any) -> Any:
    if isinstance(value, float) and pd.isna(value):
        return None
    return value.item() if hasattr(value, "item") else value


def _records(frame: pd.DataFrame) -> list[dict[str, Any]]:
    return [{str(k): _clean(v) for k, v in r.items()} for r in frame.to_dict("records")]


# --- Queue schedule ------------------------------------------------------------------------


@router.get("/queues/{queue_id}/schedule")
def queue_schedule(queue_id: str) -> dict[str, Any]:
    data = _completed()
    if data is None:
        raise HTTPException(status_code=404, detail="No schedule has been generated yet")
    queue_meta = data.meta["queues"].get(queue_id)
    shifts = data.shifts[data.shifts["queue_id"] == queue_id]
    if queue_meta is None and shifts.empty:
        raise HTTPException(status_code=404, detail=f"Queue {queue_id} is not in the schedule")
    agent_ids = set(shifts["agent_id"])
    acts = data.activities[data.activities["agent_id"].isin(agent_ids)]
    acts_by = {
        k: _records(g.sort_values("start")) for k, g in acts.groupby(["agent_id", "work_date"])
    }
    agents = []
    for agent_id, g in shifts.groupby("agent_id"):
        g = g.sort_values("work_date")
        first = g.iloc[0]
        days: list[dict[str, Any]] = []
        for r in _records(g):
            r["activities"] = [
                {k: a[k] for k in ("activity_type", "start", "end", "minutes", "paid", "queue_id")}
                for a in acts_by.get((agent_id, r["work_date"]), [])
            ]
            days.append({k: r[k] for k in (
                "work_date", "status", "shift_code", "shift_name", "shift_start", "shift_end",
                "paid_minutes", "unpaid_minutes", "preferred_shift_matched", "preferred_day_off",
                "note",
                "activities",
            )})  # fmt: skip
        agents.append(
            {
                "agent_id": agent_id,
                "agent_name": first["agent_name"],
                "paid_minutes": int(g["paid_minutes"].sum()),
                "working_days": int((g["status"] == "Working").sum()),
                "preferred_shift_matches": int((g["preferred_shift_matched"] == True).sum()),  # noqa: E712
                "days": days,
            }
        )
    coverage = data.coverage[data.coverage["queue_id"] == queue_id]
    return {
        "run_id": data.meta["run_id"],
        "created_at": data.meta["created_at"],
        "horizon": data.meta["horizon"],
        "stale": _is_stale(data.meta),
        "validation_passed": data.meta["validation_passed"],
        "queue": {"queue_id": queue_id, **(queue_meta or {"status": "not solved", "stages": []})},
        "totals": _totals(coverage, shifts),
        "coverage": _records(coverage),
        "agents": agents,
    }


def _totals(coverage: pd.DataFrame, shifts: pd.DataFrame) -> dict[str, float]:
    hours = coverage["interval_minutes"] / 60
    return {
        "required_on_phone_hours": round(float((coverage["required_on_phone"] * hours).sum()), 2),
        "scheduled_on_phone_hours": round(float((coverage["scheduled_on_phone"] * hours).sum()), 2),
        "shortage_agent_hours": round(float((coverage["shortage"] * hours).sum()), 2),
        "excess_agent_hours": round(float((coverage["excess"] * hours).sum()), 2),
        "paid_hours": round(
            float(shifts.loc[shifts["status"] == "Working", "paid_minutes"].sum() / 60), 2
        ),
        "leave_hours": round(
            float(shifts.loc[shifts["status"] == "Leave", "paid_minutes"].sum() / 60), 2
        ),
        "working_shifts": int((shifts["status"] == "Working").sum()),
        "agents": int(shifts["agent_id"].nunique()),
        "peak_required_on_phone": int(coverage["required_on_phone"].max()) if len(coverage) else 0,
    }


# --- Roll-ups (queue → MU → BU → organization) ---------------------------------------------


@router.get("/schedule/summary")
def schedule_summary(
    level: Annotated[Literal["org", "bu", "mu"], Query()], code: str = ""
) -> dict[str, Any]:
    """Totals for the children of a node, plus their sum — which equals the node's own total."""
    data = _completed()
    if data is None:
        raise HTTPException(status_code=404, detail="No schedule has been generated yet")
    shifts = data.shifts
    queue_parent = shifts.drop_duplicates("queue_id").set_index("queue_id")[["bu_id", "mu_id"]]
    cov = data.coverage.join(queue_parent, on="queue_id")
    child = {"org": "bu_id", "bu": "mu_id", "mu": "queue_id"}[level]
    if level == "bu":
        cov, shifts = cov[cov["bu_id"] == code], shifts[shifts["bu_id"] == code]
    elif level == "mu":
        cov, shifts = cov[cov["mu_id"] == code], shifts[shifts["mu_id"] == code]
    rows = [
        {"code": key, **_totals(cov[cov[child] == key], shifts[shifts[child] == key])}
        for key in sorted(set(shifts[child]))
    ]
    return {
        "run_id": data.meta["run_id"],
        "level": level,
        "code": code,
        "children": rows,
        "total": _totals(cov, shifts),
    }


# --- Agent schedule (real plan) ------------------------------------------------------------


def agent_plan(agent_id: str) -> dict[str, Any] | None:
    """The agent's weekly calendar from the latest completed run, or None if not scheduled."""
    data = _completed()
    if data is None:
        return None
    rows = data.shifts[data.shifts["agent_id"] == agent_id].sort_values("work_date")
    if rows.empty:
        return None
    acts = data.activities[data.activities["agent_id"] == agent_id]
    by_day = {d: _records(g.sort_values("start")) for d, g in acts.groupby("work_date")}
    plan = {r["work_date"]: r for r in _records(rows)}
    first = date.fromisoformat(min(plan))
    last = date.fromisoformat(max(plan))
    weeks = []
    week_start = first - timedelta(days=first.weekday())
    while week_start <= last:
        days: list[dict[str, Any]] = []
        for i in range(7):
            d = week_start + timedelta(days=i)
            r = plan.get(d.isoformat())
            if r is None:
                days.append(
                    {
                        "date": d.isoformat(),
                        "weekday": WEEKDAYS[i],
                        "status": "Outside horizon",
                        "off": True,
                    }
                )
                continue
            days.append(
                {
                    "date": d.isoformat(),
                    "weekday": WEEKDAYS[i],
                    "status": r["status"],
                    "off": r["status"] != "Working",
                    "shift_code": r["shift_code"],
                    "shift_name": r["shift_name"],
                    "start": r["shift_start"],
                    "end": r["shift_end"],
                    "paid_hours": (r["paid_minutes"] or 0) / 60,
                    "preferred_shift_matched": r["preferred_shift_matched"],
                    "preferred_day_off": r["preferred_day_off"],
                    "note": r["note"],
                    "activities": [
                        {
                            k: a[k]
                            for k in (
                                "activity_type",
                                "start",
                                "end",
                                "minutes",
                                "paid",
                                "queue_id",
                            )
                        }
                        for a in by_day.get(d.isoformat(), [])
                    ],
                }
            )
        in_horizon = [x for x in days if x["status"] != "Outside horizon"]
        weeks.append(
            {
                "week_start": week_start.isoformat(),
                "paid_hours": sum(x.get("paid_hours") or 0 for x in in_horizon),
                "days_off": sum(1 for x in in_horizon if x["status"] == "Off"),
                "leave_days": sum(1 for x in in_horizon if x["status"] == "Leave"),
                "preferred_shift_matches": sum(
                    1 for x in in_horizon if x.get("preferred_shift_matched") is True
                ),
                "working_days": sum(1 for x in in_horizon if x["status"] == "Working"),
                "partial": len(in_horizon) < 7,
                "days": days,
            }
        )
        week_start += timedelta(days=7)
    return {
        "source": "optimized",
        "run_id": data.meta["run_id"],
        "validation_status": "Validated" if data.meta["validation_passed"] else "Validation failed",
        "queue_status": data.meta["queues"].get(str(rows.iloc[0]["queue_id"]), {}).get("status"),
        "timezone": rows.iloc[0]["timezone"],
        "weeks": weeks,
    }


# --- CSV export ----------------------------------------------------------------------------


@router.get("/schedule/runs/{run_id}/export/{kind}.csv")
def export_csv(
    run_id: str,
    kind: Literal["shifts", "activities", "coverage", "requirements"],
    queue_id: str | None = None,
) -> StreamingResponse:
    data = _load(run_id)
    frame: pd.DataFrame = getattr(data, kind)
    if queue_id:
        if kind == "activities":
            frame = frame[
                frame["agent_id"].isin(
                    data.shifts.loc[data.shifts["queue_id"] == queue_id, "agent_id"]
                )
            ]
        else:
            frame = frame[frame["queue_id"] == queue_id]
    buffer = io.StringIO()
    frame.to_csv(buffer, index=False)
    name = f"schedule_{run_id}_{kind}{'_' + queue_id if queue_id else ''}.csv"
    return StreamingResponse(
        iter([buffer.getvalue()]),
        media_type="text/csv",
        headers={"Content-Disposition": f'attachment; filename="{name}"'},
    )
